src/visualizers/Grid.ts

```diff
@@ -822,7 +827,7 @@ earlier ones that use the _same_ style.)
     showNumber() {
         const currentNumberAsString = this.currentNumber.toString()
 
-        if (this.showNumbers) {
+        if (this.sketch && this.showNumbers) {
             this.sketch.fill(this.numberColor)
```

Comment:
again, this change looks kind of like a fall back check whether `this.sketch` is a thing?  Can you explain?

Reply:
Right, hopefully the above explanation made it clearer why an arbitrary member function of a Visualizer derived from P5Visualizer must now cope with the possibility that this.sketch is undefined. Another possibility for the Grid visualizer adjustment to the change would be as follows: ultimately, showNumber is called from draw (either directly or indirectly). And draw() gets a valid sketch argument. So we could instead change showNumber to take a sketch argument and just use the sketch argument, instead of dealing with this.sketch at all. Then it can assume the sketch argument is valid. We would just have to make sure the sketch argument is passed down through the call chain from draw() to showNumber().

Indeed, saying this suggests to me another global approach to the whole "sketch may not be valid" thing, which I could adjust this PR to use if you like:  the actual sketch object could be kept in a different property private to P5Visualizer, and it could instead implement a read-only this.sketch property via a getter that would always return a valid sketch object, or throw if it is not able to. But since all of the calls actually happen with valid sketches, it would never actually throw. That would mean putting back in all of the `this.` before the sketch references in draw, and removing all of the new typechecking on this.sketch, so there would be many fewer individual code changes in this PR. On the other hand, under the hood there would be lots more times when the resulting javascript is checking whether this.sketch is undefined.

A middle path would be to do this private sketch/getter thing, but in the big draw() functions that use the sketch lots of times and where I have already changed `this.sketch` to `sketch` in numerous occurrences, to just put a line `const sketch = this.sketch` at the top. Then you get the hidden typecheck only once, and you get the conciseness of writing `sketch` rather than `this.sketch`, while allowing all of the other functions that only use the sketch once or twice to just say `this.sketch` and not to have to do any explicit typechecking.

Let me know what path you want to go on this. Thanks!

Reply:
Well, what you suggest last (middle path) seems reasonable, but I don't have much expertise here.  Is the hidden typechecking a significant overhead really?  If so, then I'd like to avoid it (since it is running on everything we do inside the draw() loop). But I have no idea if it's an efficiency concern.  Ultimately, I trust you to make a decision.

Reply:
Hmmm. I am not really a javascript performance expert. But given the following considerations:

(A) We want visualizer writing to be as smooth as possible, and just being able to refer to `this.sketch` seemed to be pretty comfortable before;

(B) Any one check for undefined before actually using the sketch is presumably of microscopic cost, the only issue is if they really pile up into a very large number of redundant such checks;

(C) In many circumstances, one such check is basically unavoidable because the framework may well have decided to dispose of a previously-existing sketch, and

(D) In any one function, you can get down to just one such check by writing `const sketch = this.sketch` and then using the local sketch variable,

I am going to go ahead with this middle path. I will try to get the updated commit in ASAP and let everyone know when it is there. In the meantime I will mark this as draft.